core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala

```diff
@@ -0,0 +1,139 @@
+ * @param serializedTaskMetrics a `TaskMetrics` that is created and serialized on the driver side
+ *                              and sent to executor side.
+ * @param totalShuffleNum total shuffle number for current job.
+ *
+ * The parameters below are optional:
+ * @param jobId id of the job this task belongs to
+ * @param appId id of the app this task belongs to
+ * @param appAttemptId attempt id of the app this task belongs to
+ */
+private[spark] class ContinuousShuffleMapTask(
+    stageId: Int,
+    stageAttemptId: Int,
+    taskBinary: Broadcast[Array[Byte]],
+    partition: Partition,
+    @transient private var locs: Seq[TaskLocation],
+    localProperties: Properties,
+    serializedTaskMetrics: Array[Byte],
+    totalShuffleNum: Int,
+    jobId: Option[Int] = None,
+    appId: Option[String] = None,
+    appAttemptId: Option[String] = None)
+  extends Task[Unit](stageId, stageAttemptId, partition.index, localProperties,
+    serializedTaskMetrics, jobId, appId, appAttemptId)
+    with Logging {
+
+  /** A constructor used only in test suites. This does not require passing in an RDD. */
+  def this(partitionId: Int, totalShuffleNum: Int) {
+    this(0, 0, null, new Partition { override def index: Int = 0 }, null, new Properties,
+      null, totalShuffleNum)
+  }
+
+  @transient private val preferredLocs: Seq[TaskLocation] = {
+    if (locs == null) Nil else locs.toSet.toSeq
+  }
+
+  // TODO: Get current epoch from epoch coordinator while task restart, also epoch is Long, we
+  //       should deal with it.
+  var currentEpoch = context.getLocalProperty(SparkEnv.START_EPOCH_KEY).toInt
+
+  override def runTask(context: TaskContext): Unit = {
+    // Deserialize the RDD using the broadcast variable.
+    val threadMXBean = ManagementFactory.getThreadMXBean
+    val deserializeStartTime = System.currentTimeMillis()
+    val deserializeStartCpuTime = if (threadMXBean.isCurrentThreadCpuTimeSupported) {
+      threadMXBean.getCurrentThreadCpuTime
+    } else 0L
+    val ser = SparkEnv.get.closureSerializer.newInstance()
+    // TODO: rdd here should be a wrap of ShuffledRowRDD which never stop
+    val (rdd, dep) = ser.deserialize[(RDD[_], ShuffleDependency[_, _, _])](
+      ByteBuffer.wrap(taskBinary.value), Thread.currentThread.getContextClassLoader)
+    _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
+    _executorDeserializeCpuTime = if (threadMXBean.isCurrentThreadCpuTimeSupported) {
+      threadMXBean.getCurrentThreadCpuTime - deserializeStartCpuTime
+    } else 0L
+
+    var writer: ShuffleWriter[Any, Any] = null
+    val manager = SparkEnv.get.shuffleManager
+    val mapOutputTracker = SparkEnv.get.mapOutputTracker
+      .asInstanceOf[ContinuousMapOutputTrackerWorker]
+
+    while (!context.isCompleted() || !context.isInterrupted()) {
+      try {
+        // Create a ContinuousShuffleDependency which has new shuffleId based on continuous epoch.
```

Comment:
Who's cleaning up the old one?

Reply:
The old one(ShuffleDependency) register cleaner on driver side. ContinuousShuffleDependency we created on executor side, the tmp data delete by epoch coordinator receive commit event(Not in this PR).

Reply:
The cleanup should probably be in the same PR as the registration.

Won't that strategy leak resources if the query ends up shut down before the commit event is received?